clang/lib/3C/ProgramInfo.cpp

```diff
@@ -403,12 +405,38 @@ ProgramInfo::insertIntoExternalFunctionMap(ExternalFunctionMapType &Map,
         newC->brainTransplant(oldC, *this);
         Map[FuncName] = newC;
         RetVal = true;
-      } else
+      } else {
         // if the current FV constraint is not a definition?
         // then merge.
-        oldC->mergeDeclaration(newC, *this);
-    } else if (newC->hasBody())
-      llvm::errs() << "Warning: duplicate definition " << FuncName << "\n";
+        std::string BaseFailed = "";
+        std::string ReasonFailed = "";
+        oldC->mergeDeclaration(newC, *this, ReasonFailed);
+        bool MergingFailed = BaseFailed != ReasonFailed;
+        if (MergingFailed) {
+          clang::DiagnosticsEngine &DE = C->getDiagnostics();
+          unsigned MergeFailID = DE.getCustomDiagID(DiagnosticsEngine::Fatal,
+                                                    "merging failed for %q0 due to %1");
+          const auto Pointer = reinterpret_cast<intptr_t>(FD);
+          const auto Kind = clang::DiagnosticsEngine::ArgumentKind::ak_nameddecl;
+          auto DiagBuilder = DE.Report(FD->getLocation(), MergeFailID);
+          DiagBuilder.AddTaggedVal(Pointer, Kind);
+          DiagBuilder.AddString(ReasonFailed);
+        }
+        if(MergingFailed) {
+          // Kill the process and stop conversion
+          // Without this code here, 3C simply ignores this pair of functions
+          // and converts the rest of the files as it will (in semi-compliance
+          // with Mike's (2) listed on the original issue (#283)
+          exit(1);
+        }
+      }
+    } else if (newC->hasBody()) {
+      clang::DiagnosticsEngine &DE = C->getDiagnostics();
+      unsigned DuplicateDefinitionsID = DE.getCustomDiagID(DiagnosticsEngine::Fatal,
+                                                "duplicate definition for function %0");
+      DE.Report(FD->getLocation(), DuplicateDefinitionsID).AddString(FuncName);
+      exit(1);
```

Comment:
Instead of exiting here, just issue a warning, the way it was done before?

Reply:
Or, if we do that, perhaps that doesn't solve the `libArchive` problem (don't remember)?

Reply:
The libarchive problem will actually just be fixed via the current stuff in mergeDeclaration, so we can issue a warning here and still have things work according to plan. The reason I decided to fail here is because we decided that was the correct solution during status. 
You noted on slack that maybe failing when we see this is fine since the files won't get rewritten properly anyway. It was my understanding that rewriting doesn't suffer overly with this problem. In fact, now that you mention it, I'm fairly certain `libtiff` relies on this warning, since @aaronjeline added this stuff in when he was solving the `libtiff` crash ages ago. My understanding is that though we have these duplicate functions, we still get to correctly rewrite a bunch of other functions without incident, so I think reverting to warning is the right solution. 

But, w.r.t. question 2 below, I'll double-check how the rewriter handles this (if it were a warning) and report back, and if you agree that we should revert back to warning, then I can then add the code to make sure that the prototypes are compatible with one another.

Reply:
Sounds good. Please do try out an example with two definitions of the same function in different files and let's see how rewriting might be affected. Also see what happens if both versions have different types vs. the same types.

Reply:
Alright, tried out some wacky combinations, including functions that flow into other (unaffected, i.e. non duplicated functions) to see how rewriting performs, and I think it does the right thing from what I've seen. In the case where the prototypes are identical, both functions get converted in the exact same way, but when their types are different, they both become marked wild (in addition to the functions they flow to), but they retain their signatures pre-conversion (i.e. the resulting pair of files still compiles).

Reply:
I think we just agreed in our face-to-face meeting that we should die if there are multiple definitions. It's asking for trouble to keep going ... (just make the tests' duplication functions `static`)